Refresh the module selection after changing the course

m1r0 opened the pull request:
Fixes #3202

### Changes proposed in this Pull Request

* Refresh the module selection dropdown on the lesson edit screen after changing the course.

### Testing instructions

* Populate your database with a course with some modules and lessons and another course with only lessons (no modules).
* Go to the edit screen of a lesson.
* Select the course that has modules.
* Make sure the "module" metabox contains a dropdown with the course modules.
* Select the course that has no modules.
* Make sure the "module" metabox contains: `No modules are available for this lesson yet. Please add some to the course.`
* Select no course.
* Make sure the "module" metabox contains: `No modules are available for this lesson yet. Please select a course first.`
* Make sure the expected module value is stored in the database after updating the lesson.

### Deprecated Code
 * `Sensei_Core_Modules::ajax_get_course_modules` - in favour of
 `Sensei_Core_Modules::handle_get_lesson_module_metabox`

### Screenshot

![Screen Capture on 2021-11-15 at 21-21-45](https://user-images.githubusercontent.com/1612178/141841457-5efbbf73-9803-4e2f-a984-c96de1238c11.gif)

m1r0 commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Module selection appears only after refreshing](https://trello.com/c/JEAJSkyr/1227-module-selection-appears-only-after-refreshing)

donnapep commented on includes/class-sensei-modules.php:
```diff
@@ -205,34 +207,57 @@ public function get_lesson_module_if_exists( $post ) {
+	 * Renders the lesson module select input.
+	 *
+	 * @param int|null $course_id
+	 * @param int|null $current_module_id
```
The parameters and return type are missing the description. I often refer to the [PHP Documentation Standards](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods) when I'm unsure about commenting something or as a refresher.

m1r0 replied:
Thanks for the tip. Addressed in https://github.com/Automattic/sensei/pull/4421/commits/e46bc9d2cef52b8596bcbf2c744432bd724f349c.

donnapep commented on includes/class-sensei-modules.php:
```diff
@@ -205,34 +207,57 @@ public function get_lesson_module_if_exists( $post ) {
+				/*
+				 * translators: The placeholders are as follows:
+				 *
+				 * %1$s - <em>
```
It's actually considered to be better to include HTML directly in the string. For context, please see this [excellent resource](https://developer.wordpress.com/2015/04/23/wordpress-developers-test-your-i18n-internationalization-knowledge/) by an A11n, namely the second example. It seems like a good opportunity to correct it.

m1r0 replied:
Thanks for the info Donna, this localization wasn't done by me. I've just restructured the code. If I change it, that might break some of the translations. Should I still do it?

donnapep replied:
That's the question of the day, isn't it? 😆 It's definitely a trade-off and I'm not sure what the right answer is. On the one hand, because it's in the backend it seems like less of a big deal to change than it would be if it were something that appeared on the frontend. On the other hand, is it _really_ that big a deal? Probably not. I'm good with leaving it. 👍🏻

donnapep commented on includes/class-sensei-modules.php:
```diff
@@ -205,34 +207,57 @@ public function get_lesson_module_if_exists( $post ) {
-			 * %4$s - </a>
-			 */
-			$html .= '<p>' . wp_kses_post( sprintf( __( 'No modules are available for this lesson yet. %1$sPlease add some to %3$sthe course%4$s.%2$s', 'sensei-lms' ), '<em>', '</em>', '<a href="' . esc_url( $course_url ) . '">', '</a>' ) ) . '</p>';
+			$html .= '<input type="hidden" name="' . esc_attr( $input_name ) . '" value="">';
```
For my own clarification, what is this hidden field used for?

m1r0 replied:
This is a common strategy to make a form submit an input with no value. We need this to clear the already selected module.

Consider the following scenario:
* I select a course with modules and the module select box is updated.
* I select a module and save the post.
* I select a course with no modules and save the post. <- At this point, we need to have a `lesson_module` input with no value in order to clear the previously selected module.

Another option would be to add a hook that listens on post save and clear the module, but I find this approach simpler and easier to maintain.

donnapep reviewed (commented):
Looking good Miro! Left a few comments for when you have some time. 🙂

m1r0 commented on assets/js/modules-admin.js:
```diff
@@ -202,36 +202,28 @@ jQuery( document ).ready( function () {
-				action: 'sensei_get_course_modules',
+	// Refresh the modules meta box on course select change.
+	jQuery( '#lesson-course-options' ).on( 'change', function () {
+		const lessonId = jQuery( '#post_ID' ).val();
```
> I think I may have missed this in the other PR, but how confident are you that #post_id will always be available in the editor?

The `#post_id` element is part of the `the_block_editor_meta_boxes` core function, so I've thought it's pretty safe to use, just because it's part of blocks core. I just found another way to get the ID using `wp.data.select( 'core/editor' ).getCurrentPostId()`. Should I go with the second option?

>Also, what do you think about doing a check that both lessonId and courseId have a value before making the AJAX call? We could save this call if either is not set for some reason.

The `courseId` can be empty, so we shouldn't check its value. This is the case when the user selects "None" as a course.

I'm not a big fan of code failing silently, so that's why I'm not checking the `lessonId`. It makes bugs very difficult to find and debug. If for some reason the `lessonId` is empty, there should be an error somewhere (in the backend in this case), because this is unexpected behavior. I'd love to hear your thoughts.

donnapep replied:
> The #post_id element is part of the the_block_editor_meta_boxes core function, so I've thought it's pretty safe to use, just because it's part of blocks core. I just found another way to get the ID using wp.data.select( 'core/editor' ).getCurrentPostId(). Should I go with the second option?

Nope, if you think it's safe then that's good enough for me! 🙂 I usually see `wp_localize_script` used for this sort of thing, but this way might just be better.

> I'm not a big fan of code failing silently, so that's why I'm not checking the lessonId. It makes bugs very difficult to find and debug. If for some reason the lessonId is empty, there should be an error somewhere (in the backend in this case), because this is unexpected behavior.

m1r0 replied:
> Nope, if you think it's safe then that's good enough for me! 🙂 I usually see wp_localize_script used for this sort of thing, but this way might just be better.

Thinking about it now, maybe the safer option would be to use the `wp.data.select( 'core/editor' ).getCurrentPostId()` (the selector was created 5 years ago) just because it's a [documented API](https://developer.wordpress.org/block-editor/reference-guides/data/data-core-editor/#getcurrentpostid). I much prefer that than relying on a DOM element. I've updated the code in https://github.com/Automattic/sensei/pull/4421/commits/c4bc587967ef289f50fc0f343bb9a25d8c94ff3b. Let me know if you have any concerns.

> That makes sense. Another alternative might be to surface the missing lesson ID by logging it in the console in the frontend.

I've updated the code to use the wp data module, so I guess an error will be thrown in the frontend if that doesn't work for some reason.

donnapep replied:
Would you mind rebasing to pull in Renatho's `package.json` changes so I can rebuild more easily?

m1r0 replied:
Yep, done.

m1r0 commented on includes/class-sensei-modules.php:
```diff
@@ -2099,20 +2130,20 @@ public static function add_new_module_term() {
-	 * Get course modules
+	 * Handles the lesson module meta box ajax request by outputting the box content HTML.
 	 */
-	public function ajax_get_course_modules() {
```
Makes sense. Fixed in https://github.com/Automattic/sensei/pull/4421/commits/a50f4433bdfa67a41037aafc9969900bbb283a88.

donnapep replied:
See my comments about this [here](https://github.com/Automattic/sensei/pull/4421#commitcomment-61235342).

donnapep commented on includes/class-sensei-modules.php:
```diff
@@ -94,6 +94,7 @@ public function __construct( $file ) {
 		// store new modules created on the course edit screen
 		add_action( 'wp_ajax_sensei_add_new_module_term', array( 'Sensei_Core_Modules', 'add_new_module_term' ) );
 		add_action( 'wp_ajax_sensei_get_course_modules', array( $this, 'ajax_get_course_modules' ) );
```
Although we need to keep the deprecated function around for awhile in case others are calling it, I don't think we need to continue calling it ourselves, or do we? Otherwise we'll have deprecation notices showing up in the logs.

That also made me wonder if there's any more code to remove related to the first parameter. I went looking for other instances of `sensei_get_course_modules` but didn't find any, which made me wonder if this hook every actually did anything in the first place. 🤔 

One final step in the deprecation process is to add the deprecated code to an issue that is attached to the milestone in which it will be removed. Based on our deprecation policy, that will be in 5.0. There's already an [issue](https://github.com/Automattic/sensei/issues/2976) created for that, so you can add your's there.

m1r0 replied:
> Although we need to keep the deprecated function around for awhile in case others are calling it, I don't think we need to continue calling it ourselves, or do we?

I've left the ajax hook just in case someone else is making requests to it. Otherwise, we are not calling it anymore. There was some old JS code calling the action, but I've replaced it with the new action. Let me know if I should remove it.

>That also made me wonder if there's any more code to remove related to the first parameter. I went looking for other instances of sensei_get_course_modules but didn't find any, which made me wonder if this hook every actually did anything in the first place. 🤔

There was some code [that I've removed](https://github.com/Automattic/sensei/pull/4421/files#diff-99b5acd02bfdd8020b116bd2c4a896a110dc852033844c6d4c1a4c120b8f3be2L219) which was calling the hook, but for some reason, that code was not working. Basically, you are correct - the hook didn't do anything. Maybe it was working before the blocks were introduced.

>One final step in the deprecation process is to add the deprecated code to an issue that is attached to the milestone in which it will be removed.

Done, thanks!

donnapep replied:
> I've left the ajax hook just in case someone else is making requests to it.

Good call! I wonder if we should leave a note in #2976 so we don't forget to remove the associated hook when we remove the function?

m1r0 replied:
Makes sense. Done!

donnapep reviewed (dismissed):
💯

donnapep commented:
@m1r0 So I just thought to use this with the Classic Editor plugin, and it doesn't work there. Although we don't officially support the classic editor, I'm not sure I'm comfortable with breaking things for those who are using it. Is there a solution that works in both editors?

m1r0 commented:
> So I just thought to use this with the Classic Editor plugin, and it doesn't work there. Although we don't officially support the classic editor, I'm not sure I'm comfortable with breaking things for those who are using it. Is there a solution that works in both editors?

@donnapep Managed to do a quick fix by using both methods (getting the id from the wp data store + the DOM as a fallback). Fixed in b0a1ba5fc7ce72fea540fc092f757f8dbdc04abf. Also, another issue came up, the blocks toggling control was causing some troubles when the block editor is disabled. Fixed in 6a8d23b4f7a82a3a5ea533efb4cf4a8bcaf48b33.

donnapep reviewed (approved):
🥳
